### ISSUE TITLE
Mongoose slows down performance of VSCode (TSServer) by a lot

### ISSUE THREAD

**Issue by ShadiestGoat** [Automattic/mongoose#10349]
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

It slows down vscode by a lot, even with small projects. I have a very small project, with like 2 small schemas, and the 2 other files, containing on average 160 lines of code, and vscode slows down a lot as soon as I install mongoose. This does not happen when I install most other libraries (haven't tested them all lmao). If I uninstall mongoose, it comes back and works perfectly normal. 
To clarify "slowing down" means that typescript language server needs 3-5 seconds to realise that I'm importing somethign invalid, and just vscode in general to realise I'm importing something that's never used takes around 2-4 seconds. Usually it happens in under a second so like.. you know...
And it's not like vscode is taking a lot of ram/cpu either, it just grows slow. 
Btw I think this is a mongoose issue, since it happens with no other library, its a special case here
Reloading typescript server or project doesn't help

**If the current behavior is a bug, please provide the steps to reproduce.**

Have vscode to edit your code
Install mongoose on your current project (`npm i mongoose` 

My tsconfig.json:
```json
{
    "compilerOptions": {
        "allowJs": true,
        "moduleResolution": "Node",
        "esModuleInterop": true,
        "target": "ES6",
        "resolveJsonModule": true,
        "outDir": "./dist",
        "checkJs": true,
        "module": "CommonJS",
        "lib": ["ESNext"]
    },
    "include": ["src"]
}
```
**What is the expected behavior?**

It to not be like this, lmao.

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**

Node: v16.2.0
Mongoose: 5.12.13
Linux: 5.12.7-arch1-1

**Comment by vkarpov15** [comment]
Interesting, can you please provide an example project that demonstrates this? I haven't noticed this but I might be missing something.

**Comment by ShadiestGoat** [comment]
Mm well I can replicate it with very little:
```ts
import { model, Model, Schema } from "mongoose";

export interface UserSH {
    username: string,
    password: string,
    maxMb: number,
    submitted: string[],
    id: string
}

export const UserSchema = new Schema<UserSH, Model<UserSH>, UserSH>({
    id: "string",
    maxMb: "number",
    username: "string",
    submitted: ["string"],
    password: "string"
})


uiyasdjklsadias //error this baby is undefined
```
The thing that 'triggers' it is creating the schema. If I just remove the schema creation, its all good

**Comment by mjfwebb** [comment]
I can confirm that I am also experiencing a degradation in Typescript performance due to Mongoose.

**Comment by jasonwwl** [comment]
me too

**Comment by cainaf** [comment]
I can confirm that too.

I was using @types/mongoose before and just removed it in favor of mongoose own type definitions. It was like snapping two fingers, boom, vscode was now super slow. Simple autocompletes and validations are taking now over 5 or even 10 seconds sometimes in all project files.

**Comment by xiaoxin-sky** [comment]
我也能够明显感觉到，mongoose 使得 vscode 加载类型提示变慢。但是仅限于加载 mongoose 相关的内容变慢。

**Comment by ShadiestGoat** [comment]
Really? For me it slows all content, not just mongoose related...

**Comment by cainaf** [comment]
Same as @ShadiestGoat for me. Even `eslint` gets very slow. If I delete `index.d.ts` from mongoose lib inside node_modules, performance of everything becomes instant again.

**Comment by FernandoCervantesG** [comment]
I can confirm it slows down all the content for me too.

I did as @cainaf said to verify it was mongoose, I deleted `index.d.ts` and..., vs code was running fine as hell again.
I restored index.d.ts and the performance heavily dropped down again.

**Comment by ShadiestGoat** [comment]
Well removing `new Schema<UserSH, Model<UserSH>, UserSH>` removes the issues with performance (on the script I gave above)

**Comment by otress** [comment]
I am experiencing the problem since Mongoose has adopted the official type definition. I have a workaround that is to delete index.d.ts of mongoose in node_module and then install @types/mongoose@5.10.5

**Comment by xiaoxin-sky** [comment]
有什么进展吗？

**Comment by TigranAvagyan18** [comment]
Experiencing same problem . Any solutions to this?

**Comment by qafoori** [comment]
And I have the same problem in vscode. I was using postgresql in my project and for some reason I migrated to mongodb (using mongoose). vscode has been super slow since I migrated to mongodb.

**Comment by TigranAvagyan18** [comment]
> I am experiencing the problem since Mongoose has adopted it's own official type definition. I have a workaround that is to delete `index.d.ts` of mongoose in node_module and then `npm i -D -E @types/mongoose@5.10.5` to temporarily use the DefinitelyTyped type definition.

Did u get any performance change after performing that?

**Comment by nikibiki61** [comment]
This really needs a fix... Programming is not possible like this.

**Comment by ShadiestGoat** [comment]
So I'm looking into potential reasons for this (just started). To me it just looks like its a long build of the problem, where the types file is just too long/complex. For example, taking an older version of mongoose (eg. 3.11.18), the issue was starting to present itself, given that it decreased performance by a lot, but no where near this level. 
It could be that to solve this problem the types for this module need a major rework. I also believe this should be a priority.

**Comment by ShadiestGoat** [comment]
I'm not sure how much it would help, but perhaps splitting the types into multiple files may help? I'm pretty sure VSCode uses tsserver, which (at least I'm pretty sure of which) has to load an entire file at once

**Comment by ShadiestGoat** [comment]
Oh and forgot to mention this earlier, but, I have a pretty good laptop I'm coding on. HP Omen 15, with ryzen 7 4800H, SSD, and a gpu, so it isn't like I just gotta get a better laptop

**Comment by Thisura98** [comment]
I created a new TypeScript Node project and it worked fine even just after installing mongoose. However, as soon as I started creating a schema and creating a model from it, things started going south. 

This is what the schema + model code looks like.

```ts
import * as mongoose from 'mongoose';

export const testSchema = new mongoose.Schema({
    name: String,
    time: Number,
    age: Number
});

export const testModel = mongoose.model('TestModel', testSchema);
```

I also checked by importing the `model` and `Schema` individually instead of using `* as` but the outcome was the same.

I am a beginner to the world of Mongo and Node so unfortunately I am not knowledgable enough to debug this issue.  So instead for the time being, I am going to resort to using the vanilla MongoDB Driver for Node: [link](https://docs.mongodb.com/drivers/node/current/quick-start/).

Sorry mongoose.

**Comment by Thisura98** [comment]
Decided to give it one more shot by taking into consideration feedback by @TigranAvagyan18. __And I got it to work!__

Commands:

```
npm i mongoose@5.10.19 --save
npm i @types/mongoose@5.10.5 --save
```

Summary:

- Downgrade to last version before Mongoose's own TS types
- Install DefinitelyTyped's Mongoose types.

My existing code worked with these versions without issues. As a side note, Mongoose introduced their own types in version 5.11.0 [according to this](https://mongoosejs.com/docs/typescript.html). 5.10.19 was the last version released before that [according to this](https://www.npmjs.com/package/mongoose?activeTab=versions)

I will test some more tonight to see if my current code was affected by this downgrade and update this post if I find any.

**Comment by cainaf** [comment]
Although returning to `@types/mongoose` does solve performance problem, it also may break your code. It sure did mine.

When we migrated to mongoose own types, we had to fix more than 500 typescript errors (the code was fine and would work and required no change, it's just the way types are implement that vary from one to another, the requirement of methods and their return types).

If we were to fallback to `@types/mongoose` again, that's what's waiting us:

![Captura de Tela 2021-07-23 às 14 27 34](https://user-images.githubusercontent.com/2081275/126819618-7cb6c5ef-d97a-41ba-83cb-9861e32c4cf0.png)

And then, when the problem here is solved, we would have to change all over again. not cool  =/

**Comment by ShadiestGoat** [comment]
Using `@types/mongoose` is a hasckish workaround, that may cause a lot of issues. It is not a solution.

**Comment by haouarihk** [comment]
my guess is that tsserver gets in an infinite loop before breaking out in few seconds, i'm currently searching of any type loops in the ``index.d.ts``
can happen recursionly

**Comment by ShadiestGoat** [comment]
@haouarihk could it be that `SchemaTypeOptions` has `Schema` in it, which has `SchemaTypeOptions`?

**Comment by haouarihk** [comment]
i'm not an expert at typescript, i can read it. But i can't identify the problem quickly. specially dealing with 3k lines of code.

@ShadiestGoat that's where i was looking, but i can't do more.

**Comment by ShadiestGoat** [comment]
Hm actually, I've been experimenting with just commenting and uncommeting large amounts of code in the declaration file. 
I have narrowed it down to stuff in the `Model` interface! I will take a closer look at it after some lunch haha

**Comment by eterner0423** [comment]
Files associated with mongoose are slow, and there are many union types in the declaration document.

**Comment by ShadiestGoat** [comment]
However the unions don't make everything the 4-5 seconds wait that the contents of the Model interface do

**Comment by cainaf** [comment]
@ShadiestGoat i did as you experiment and got a little performance improvement commenting out either `interface Model` or `class Schema`, but greatness came after commenting out them both together

**Comment by ShadiestGoat** [comment]
I've looked into further details, and while this could be just a coincidence, but whenever uncommenting anything that uses one of the generics, that when it gets slow.

**Comment by cainaf** [comment]
commenting out line 2440 alone has given me MAJOR performance improvement:

```ts
type _AllowStringsForIds<T> = {
  // [K in keyof T]: [Extract<T[K], mongodb.ObjectId>] extends [never] ? T[K] : T[K] | string;
};
```

and it doesn't seem to break my code. it's not perfect, but at least I feel like I can work again now

**Comment by ShadiestGoat** [comment]
I don't see why thatd be it... It's mentioned only once. Can @vkarpov15 explain why `LeanDocument` is made the way that it is? I honestly don't get it. To me it just looks like 
```ts
type LeanDocument<T> = {
    _id?: T;
    __v?: any;
    id?: any;
}
```
Would do the trick. I don't get exactly what LeanDocument is going for

**Comment by vkarpov15** [comment]
@ShadiestGoat `LeanDocument` is for pulling out the type returned by `toObject()`, `toJSON()`, and `lean` queries, in case the user gives us a class that `extends Document`. We've considered getting rid of it because I suspect it may have something to do with the performance issues you're seeing - can you confirm that getting rid of `LeanDocument` improves performance in your case?

**Comment by ShadiestGoat** [comment]
I don't think so. The declarations document I think has *multiple* cuases of this issue. I don't see much of a difference when changing `LeanDocument<T>` to be `T`, although different users will see it as different

Again as stated before I see major issues with the `Model` interface. I am having big lag spikes when uncommenting anything from the `Model` interface that uses the generic type `T`

I think this may need a different generics system, possibly recoding from scratch the declarations for the Schema, with more than likely backwards breaking changes

**Comment by andreialecu** [comment]
I created a minimal repro at:
https://github.com/andreialecu/repro-mongoose-slow-ts

I'm on a M1 Apple CPU so I'm not super affected by this, but I did notice a minor slowdown.

There's an easy process to verify it:

- run `yarn tsc --extendedDiagnostics` in the root. I'm seeing `Check time: 1.11s`
- If I comment out the `new Schema` call, the check time is `0s`.
- If I comment out the line mentioned in https://github.com/Automattic/mongoose/issues/10349#issuecomment-885927142 the check time drops to 0.6s

**Comment by perenSHERE** [comment]
> **It also affects the performance in a javascript application too** 

So I have created a minimal repro **for javascript** at:
https://github.com/perenSHERE/mongoose-repro-intellicode-javascript
There's an easy process to verify it:

- if you comment the whole exampleSchema const you will see a good performance boost in  the intellicode
- if you then uncomment it then try it makes the intellicode really laggy again

**Comment by andreialecu** [comment]
> I'm on a M1 Apple CPU so I'm not super affected by this, but I did notice a minor slowdown.

I retract that, there's a pretty major and very annoying slowdown in actual usage even with `skipLibCheck`

I've been digging through the types and I noticed that the biggest slowdown is caused by the interaction between `FilterQuery<T>` and `DocumentDefinition<T>`.

If I replace FilterQuery with this:

```ts
  type _FilterQuery<T> = {
    [P in keyof T]?: P extends '_id'
    ? [Extract<T[P], mongodb.ObjectId>] extends [never]
    ? mongodb.Condition<T[P]>
    : mongodb.Condition<T[P] | string | { _id: mongodb.ObjectId }>
    : [Extract<T[P], mongodb.ObjectId>] extends [never]
    ? mongodb.Condition<T[P]>
    : mongodb.Condition<T[P] | string>;
  } &
    mongodb.RootQuerySelector<DocumentDefinition<T>>;

  export type FilterQuery<T> = _FilterQuery<T>;
```

The check time drops by 30-40% via `tsc --extendedDiagnostics`. It also seems to cause no problems in my code base and I get most of the speed back.

I opened a PR at https://github.com/Automattic/mongoose/pull/10515

**Comment by perenSHERE** [comment]
> **This PR might be able to fix some issues but not issues with the systems who have half decent specification**

**my system specifications are:**
- i5 2500k 
- 8 GB ram
- gt 1030

**before applying changes in the PR #10515**
```
Files:                          87
Lines of Library:            28119
Lines of Definitions:        25937
Lines of TypeScript:            19
Lines of JavaScript:             0
Lines of JSON:                   0
Lines of Other:                  0
Nodes of Library:           119708
Nodes of Definitions:       108811
Nodes of TypeScript:            65
Nodes of JavaScript:             0
Nodes of JSON:                   0
Nodes of Other:                  0
Identifiers:                 81545
Symbols:                    128399
Types:                       43073
Instantiations:             291774
Memory used:               149894K
Assignability cache size:    13123
Identity cache size:             3
Subtype cache size:              0
Strict subtype cache size:       0
I/O Read time:               0.02s
Parse time:                  0.64s
ResolveModule time:          0.07s
ResolveTypeReference time:   0.01s
Program time:                0.76s
Bind time:                   0.36s
Check time:                  1.87s
transformTime time:          0.01s
commentTime time:            0.00s
I/O Write time:              0.00s
printTime time:              0.02s
Emit time:                   0.02s
Total time:                  3.02s
```
**after applying changes in the PR #10515**
```
Files:                          87
Lines of Library:            28119
Lines of Definitions:        25937
Lines of TypeScript:            19
Lines of JavaScript:             0
Lines of JSON:                   0
Lines of Other:                  0
Nodes of Library:           119708
Nodes of Definitions:       108811
Nodes of TypeScript:            65
Nodes of JavaScript:             0
Nodes of JSON:                   0
Nodes of Other:                  0
Identifiers:                 81545
Symbols:                     98716
Types:                       23615
Instantiations:             153096
Memory used:               125551K
Assignability cache size:     7426
Identity cache size:             1
Subtype cache size:              0
Strict subtype cache size:       0
I/O Read time:               0.02s
Parse time:                  0.63s
ResolveModule time:          0.06s
ResolveTypeReference time:   0.01s
Program time:                0.74s
Bind time:                   0.36s
Check time:                  1.40s
transformTime time:          0.01s
commentTime time:            0.00s
I/O Write time:              0.00s
printTime time:              0.02s
Emit time:                   0.02s
Total time:                  2.51s
```
> **This PR gives a bit more boost in the performance but its still not enough**

**Comment by ShadiestGoat** [comment]
It looks like using the 'raw' version of the schema (`UserSH` in my example) in as many places as possible rather than the full document increases performance, which is to be expected. I am working on some more changes of this nature, for the 6.0 release

**Comment by andreialecu** [comment]
@ShadiestGoat the problem is some sort of infinite recursivity between Schema, Model and Query, in relation to `extends Model<...>` checks that TypeScript needs to do.

Take a look at my investigation in https://github.com/Automattic/mongoose/pull/10515 and the details I added in: https://github.com/microsoft/TypeScript/issues/45249 - in particular to the performance trace there.

I also ran `tsc --generateTrace outdir` and in the resulting trace it appears that it goes up to a high depth of recursion (before probably? giving up):

<img width="409" alt="Screenshot 2021-07-30 at 18 15 58" src="https://user-images.githubusercontent.com/697707/127674398-7a55661c-31ba-4364-949e-ad891b716898.png">

<img width="211" alt="Screenshot 2021-07-30 at 18 16 08" src="https://user-images.githubusercontent.com/697707/127674400-a542765d-5c87-482b-a7a3-6279b3d697f2.png">

Using the `raw` version of the schema doesn't help too much, the repro I attached in a previous comment uses your example explicitly but still exhibits the problem.

**Comment by andreialecu** [comment]
I've been randomly getting this from typescript during compilation:

```
../node_modules/mongoose/index.d.ts(706,5): error TS2502: 'schema' is referenced directly or indirectly in its own type annotation.
```

Offending lines:
https://github.com/Automattic/mongoose/blob/9b00c96d442565e99bb5313f2426cc81758bb804/index.d.ts#L705-L706


It doesn't always happen, I guess it's related to the recursion counter sometimes going over a threshold.

**Comment by ShadiestGoat** [comment]
In there, shouldn't that be the 'raw' version of the schema? Idk

**Comment by vkarpov15** [comment]
We made some improvements in fefebb3 that will be shipped with v5.13.7, and we'll continue trying to find other ways to speed this up.

**Comment by smolinari** [comment]
What was changed in the versions released up until shortly before June 13th? Because, this performance drop wasn't noticed at first with the introduction of Mongoose's own types. So, I'd venture to say, the issue was introduced in a change made shortly before June 13th. I mean, between November of last year, when the new types were introduced, until May (6 months) nobody had such issues, right? Just saying and trying to help. :)

Scott

**Comment by mjfwebb** [comment]
@smolinari anecdotal only but I have been experiencing the slowdown ever since moving to use mongoose types from 5.12.9.

It might not be that surprising then that in the comparison (https://github.com/Automattic/mongoose/compare/5.12.9...master) there sits `allow using type: new Schema({ ... })` which seems to be the biggest culprit.

**Comment by ShadiestGoat** [comment]
Hey I'm not sure if anyone else agrees with this, but the major performance issues have been solved latest version (5.13.7)?

**Comment by xiaoxin-sky** [comment]
@ShadiestGoat 性能得到了提升，但是还不够。但是在我的 vscode 上，依然能够感受到提示不及时的问题。

**Comment by lorand-horvath** [comment]
I confirm the very slow autocomplete/intellisense of VSCode when using Mongoose 5.13.5 to 5.13.7.
Initially I thought it was caused by the latest VSCode 1.59 and opened/commented in issues #130807 #129099 #130010
Is there anything I can test/try with Mongoose?

**Comment by vkarpov15** [comment]
@lorand-horvath can you come up with an example similar to the one in #10515 (https://github.com/andreialecu/repro-mongoose-slow-ts) that demonstrates the slowdown you're seeing using `tsc --extendedDiagnostics`? Seeing what the code that's causing the slowdown would help figure out where to look.

**Comment by lorand-horvath** [comment]
@vkarpov15 There is no need to go that far to show the issue. In fact, you can test it very quickly. Initialize a new npm package with npm init and install npm install mongoose@5
Then open a new file index.js in VSCode and add the following product schema & model:

`const mongoose = require('mongoose');

const productSchema = new mongoose.Schema({
    name: {
        type: String,
        trim: true,
        required: [true, 'A product must have a name'],
        unique: true
    },
    color: {
        type: String,
        required: [true, 'A product must have a color']
    },
    weight: {
        type: Number,
        required: [true, 'A product must have a weight']
    },
    ratingsAvg: {
        type: Number,
        default: 4.5
    },
    ratingsQuantity: {
        type: Number,
        default: 0
    },
    price: {
        type: Number,
        required: [true, 'A product must have a price']
    },
    priceDiscount: Number,
    summary: {
        type: String,
        trim: true,
        required: [true, 'A product must have a description']
    },
    description: {
        type: String,
        trim: true
    },
    image: {
        type: String,
        required: [true, 'A product must have an image']
    },
    images: [String],
    createdAt: {
        type: Date,
        default: Date.now(),
        select: false
    }
});

const Product = mongoose.model('Product', productSchema);

module.exports = Product;
`
Then just hover with the mouse over productSchema in the declaration and intellisense will show "Loading..." for several seconds while the CPU process blasts up to 100% CPU usage. With the latest VSCode 1.59.1 and mongoose 5.13.8 I have to wait about 5 seconds (sometimes more) every time I make any changes to the file or if I hover over one of the variables, in order to see the autocomplete results.

**Comment by lorand-horvath** [comment]
I've just tested Mongoose 6.0.0 and VSCode intellisense/autocompletes seem to be a little bit faster...

**Comment by vkarpov15** [comment]
@lorand-horvath that's good news! Might be because we removed some legacy type aliases. One of those types used `extends Document` in a generic, which is similar to the issue that was fixed by #10515, so I can see that being possible.

**Comment by traverse1984** [comment]
Can confirm that I'm also experience a big slowdown of my editor experience using Mongoose 6.0.4. I have gone with the suggestion from above to use an earlier version of Mongoose and the DT types, which doesn't cause the same issue. I would much rather be using the latest version.

**Comment by lorand-horvath** [comment]
> 
> 
> Can confirm that I'm also experience a big slowdown of my editor experience using Mongoose 6.0.4. I have gone with the suggestion from above to use an earlier version of Mongoose and the DT types, which doesn't cause the same issue. I would much rather be using the latest version.

Please tell us which exact earlier version of Mongoose are you using that doesn't cause the VSCode slowdown?

I've just tried latest VSCode 1.60 with full TS 4.4 inline hints support and bracket pair colorization, but the slowdown and high CPU usage issue still remains (both Mongoose 5.13.8 and 6.0.4). As I mentioned before in https://github.com/Automattic/mongoose/issues/10349#issuecomment-905350985 v6 is a bit faster though.

**Comment by traverse1984** [comment]
As per the comment from @Thisura98, I installed:

```
npm i mongoose@5.10.19 --save
npm i @types/mongoose@5.10.5 --save
```

**Comment by sanmmm** [comment]
It still very slow when I upgrade to version 6.0.6

**Comment by vkarpov15** [comment]
@sanmmm that's hard to believe because 6.0.6 hasn't been released yet.

On the bright side, we did some digging and found that we can repro significant perf degradation on https://github.com/andreialecu/repro-mongoose-slow-ts if we disable TypeScript's `skipLibCheck`. With Mongoose 6.0.5, we're getting 4.1 seconds and 241k instantiations with `tsc --extendedDiagnostics`. With e0fb110, we're getting 3.6 seconds and 166k instantiations. Still not all the way to where we want to be, but should help somewhat.

We'll release 6.0.6 soon, please try it out and see if that makes things faster. We'll keep iterating for future releases to see if we can speed things up more.

**Comment by sanmmm** [comment]
@vkarpov15  oh, sorry! it's a typo. The real version i have tested is 6.0.5.

**Comment by Josephvs96** [comment]
One interesting behavior i noticed when i was on 6.0.5 is when i use the new keyword to initialize a new instance of schema is when the slowness started for me. When i removed the new keyword it would fix the problem, i would lose IntelliSense on the schema object itself but hey that was trade i was willing to make. 

I tested 6.0.6 now and for me updating solved the speed issues i was having when creating a new schema with the new keyword.

**Comment by vkarpov15** [comment]
@Josephvs96 I think that's because our TypeScript definitions currently don't support using `Schema()` without `new`. So that provides some insight into where the slowness may be coming from.

@sanmmm @lorand-horvath can you please confirm whether 6.0.6 has sped things up for you?

**Comment by lorand-horvath** [comment]
@vkarpov15 I have tested Mongoose 6.0.6 with the latest VSCode 1.60.1 and I am happy to report that the problem of slow intellisense/autocomplete is now completely solved!

For reference, in my controllers.js file, Mongoose 5.13.8 takes about 8 seconds to comment out a line (until it turns green) or for a hint to appear on hover, while one CPU thread is blasting at 100% for an additional few more seconds!
Mongoose 6.0.6, on the other hand, takes about 1 second to do that and the CPU settles down soon afterwards.
So thank you for fixing this annoying ssue and sorry for nagging at you because of it.

**Comment by vkarpov15** [comment]
Thanks for your patience! We'll do some more work to see if we can further reduce the unnecessary overhead.

**Comment by lorand-horvath** [comment]
Mongoose 6.0.7 VSCode speed is great as well - nice!

**Comment by dietpizza** [comment]
I've tested the 6.0.7 release in typescript, if you dont pass the type parameter to new Schema<T>() everything becomes instant again.

**Comment by lorand-horvath** [comment]
> 
> I've tested the 6.0.7 release in typescript, if you dont pass the type parameter to new Schema() everything becomes instant again.
Obviously - there's no type checking involved.

**Comment by arnav7633** [comment]
I am having this issue. In the latest version ;(
```
import { Schema, model } from "mongoose";
import { Macros } from "../utility/interfaces";

const schema = new Schema<Macros>({
  _id: { type: String, required: true },
  macros: { type: [String], required: true },
});

const MacroModel = model<Macros>("macros", schema);

export default MacroModel;
```

**Comment by dietpizza** [comment]
@arnav7633 can you remove the type from Schema? In my case if i remove the type parameter its all good.

**Comment by lorand-horvath** [comment]
@dietpizza Removing the type from Schema defeats the whole purpose of Mongoose...
@arnav7633 It works for me as I mentioned above with versions 6.0.6 and 6.0.7. What version are you using? "latest" is not a version.

**Comment by arnav7633** [comment]
> @arnav7633 can you remove the type from Schema? In my case if i remove the type parameter its all good.
> Also it's good practice to define the type in the same file rather than in utils

That does fix it. But then i cant access queried docs props

**Comment by dietpizza** [comment]
@arnav7633 you can. You're passing it to the model method right? It shows up in thr autocomplete too

**Comment by dietpizza** [comment]
@lorand-horvath No it doesn't.

**Comment by arnav7633** [comment]
> @arnav7633 you can. You're passing it to the model method right? It shows up in thr autocomplete too

If i pass the interface to either the new Schema or model() it doubles the compile time

```
import { Schema, model } from "mongoose";
import { Document } from "mongoose";

interface Macros extends Document {
  macros: Array<object>;
}

const schema: Schema = new Schema({
  _id: { type: Schema.Types.ObjectId, required: true },
  macros: { type: [Object], required: true },
});

const MacroModel = model<Macros>("macros", schema);

export default MacroModel;
```
const MacroModel = model<Macros>("macros", schema);

export default MacroModel;
```

**Comment by dietpizza** [comment]
@arnav7633 compile time? Does your auto complete slow down?

**Comment by arnav7633** [comment]
@dietpizza Yep like by a lot! Normally i have a compile time of 6 secs, but with the interface it bumps up to like 15 secs. The autocompleted also follows the same regime. SOmetimes forcing me to restart to fix the stupid errors because its stuck on my half typed code

**Comment by dietpizza** [comment]
@arnav7633 Thats really weird. My auto complete speeds up to instant. Are you using VScode?

**Comment by arnav7633** [comment]
@dietpizza Yea i am

**Comment by lorand-horvath** [comment]
> 
> 
> @lorand-horvath No it doesn't.

Can you be more telegraphic, please?

**Comment by dietpizza** [comment]
@lorand-horvath do you use types in js? Doesn't mongoose work on js? See its simple.

**Comment by dietpizza** [comment]
@arnav7633 Me too, but it works for me just fine.

**Comment by arnav7633** [comment]
So any idea?

**Comment by arnav7633** [comment]
I get this error if i want to access document props if i dont pass the interface to the model method
```Property 'macros' does not exist on type 'Document<any, any, unknown> & { _id: ObjectId; }'.```

**Comment by dietpizza** [comment]
@arnav7633 dont extend Document.

**Comment by arnav7633** [comment]
Damn, it literally halved the compile time

**Comment by dietpizza** [comment]
@arnav7633 So back to initial compile time?

**Comment by dietpizza** [comment]
@arnav7633 Dude you don't need to worry about compile time. Just use ts-node... Or tsc --watch. Incremental compilation is a hell lot faster.

**Comment by arnav7633** [comment]
@ShadiestGoat yep all  good now. My real time type checks are directly proptional to my compile time.

**Comment by dietpizza** [comment]
@arnav7633 nice to hear that. Btw are you using ts-node?

**Comment by arnav7633** [comment]
@dietpizza no, I compile my code and run it with node

**Comment by dietpizza** [comment]
@arnav7633 do you do that for every change? Just use nodemon man 😂. Install ts-node, nodemon and run nodemon app.ts

**Comment by arnav7633** [comment]
@dietpizza nah I have a custom start script which handles the compiling and running with the watch mode on

**Comment by dietpizza** [comment]
@arnav7633  okay 😅

**Comment by lorand-horvath** [comment]
> 
> 
> @lorand-horvath do you use types in js? Doesn't mongoose work on js? See its simple.

Take it down a notch with the boasting, alright?

**Comment by lorand-horvath** [comment]
> 
> 
> do you use types in js? Doesn't mongoose work on js? See its simple.

Take it down a notch with the boasting, will ya, genius? It's that simple.

**Comment by lorand-horvath** [comment]
> 
> 
> @lorand-horvath do you use types in js? Doesn't mongoose work on js? See its simple.

@dietpizza ok, genius.

**Comment by ShadiestGoat** [comment]
Heya! I haven't really been looking at this issue since I just made my own module haha, am I right to say that this issue is resolved & should be closed?

**Comment by dietpizza** [comment]
Probably. Can a comment be marked or something that would help.

**Comment by traverse1984** [comment]
I tried 6.0.7 earlier and found things to still be pretty slow. Not as bad as it was, I don't think, but bad enough that I'd stick to the 5x release with the DT types. Is there a test case which includes typings which should work that I could run - to rule out I'm not doing it wrong?

**Comment by dietpizza** [comment]
Dude all i know is if you don't extend Document and dont pass the interface/type to Schema() my auto complete and compilation is fast... Like regular speed.

**Comment by lorand-horvath** [comment]
@ShadiestGoat It works fine for me, you can close the issue.

**Comment by traverse1984** [comment]
> Dude all i know is if you don't extend Document and dont pass the interface/type to Schema() my auto complete and compilation is fast... Like regular speed.

As said previously… doesn’t that mean you get no type information?

**Comment by lorand-horvath** [comment]
> 
> 
> > Dude all i know is if you don't extend Document and dont pass the interface/type to Schema() my auto complete and compilation is fast... Like regular speed.
> 
> As said previously… doesn’t that mean you get no type information?

@traverse1984 Yup, he just won't acknowledge it and IMHO it doesn't really make sense to use it like that. If it does make sense, please, someone genius, enlighten me.

**Comment by traverse1984** [comment]
> > > Dude all i know is if you don't extend Document and dont pass the interface/type to Schema() my auto complete and compilation is fast... Like regular speed.
> > 
> > 
> > As said previously… doesn’t that mean you get no type information?
> 
> @traverse1984 Yup, he just won't acknowledge it and IMHO it doesn't really make sense to use it like that. If it does make sense, please, someone genius, enlighten me.

I used Mongoose before I started using TS. I still think Mongoose is useful without the type information, but I would say if you don't care about that then this issue is not relevant to you. For those of us that want the typings working, "JavaScript doesn't have types" is not really a solution.

**Comment by ShadiestGoat** [comment]
Yeah I didn't read into it yet. That solution is absolutely horrible and damaging. I mean it goes against all the work put into the types! 
As a typescript user myself, if mongoose's support for types comes at a cost of slowing the tsserver down severly, it doesn't support types. This issue shall remain open until the performance is types is ok.

**Comment by dietpizza** [comment]
@traverse1984 I do get type information, as its passed to the model<Type>(). It shows up in the completion.
@lorand-horvath Dude, why so edgy?

**Comment by traverse1984** [comment]
I made a gist: https://gist.github.com/traverse1984/ee57740a87374188851b30b4286a47fa

- When you provide a schema type to `new Schema` and `model`, things are very slow as have been described.
- When you provide the schema and model types to `new Schema` and `model`, things speed back up. 
- If the model type passed to `new Schema` does not match that passed to model, it seems to slow down again. Granted, you wouldn't want to do this - but perhaps it helps?

So could the workaround for now just be to always explicitly provide the model, even if you don't need additional methods on the model? Does this happen for others?

```
const schema = new Schema<User, Model<User>>(...);
const UserModel = model<User, Model<User>>('User', schema);
```

**Comment by traverse1984** [comment]
I believe the issue may be line 103 of index.d.ts.

Change:
```
export function model<T>(name: string, schema?: Schema<any>, collection?: string, skipInit?: boolean): Model<T>;
```
to
```
export function model<T>(name: string, schema?: Schema<T>, collection?: string, skipInit?: boolean): Model<T>;
```

and the issue seems to go away for me. I'm not fully up to speed on the entire typing system in use here - so I'm not sure if there are implications to this, but the examples I provided in the gist both work with this change.

**Comment by ShadiestGoat** [comment]
> I believe the issue may be line 103 of index.d.ts.

Can anyone else confirm this? I'm away from my setup atm. @traverse1984 if this is actually a _the_ solution, please make a pr... Ill take a look at why this happens and see if anything is wrong

> should clarify that type info is not available in the schema creation but it's available when using the model. Thats better than a 3 second delay for the for the whole project

Ie. This is a workaround, not a solution.

**Comment by traverse1984** [comment]
Made a pull request. I can't see any issues with this, and in fact when you create a Schema you already specify that same type parameter on the schema - so I don't see why it would cause issues. I've included the notes in the PR, but to summarise here:

> It seems that mismatched Schema/Model types cause the slowdown. I believe (but am not certain), that the `Schema<any>` in the [model] function parameter was causing this to happen, despite the return type being correct. I noted that the Schema interface creates a model with 'any' types, but the defaults for the Model interface generics are {}.

**Comment by slevy85** [comment]
A solution while wainting for the PR #10801 to be merged is to use the second declaration of `model` function, inline 106 of index.d.ts :
```
 export function model<T, U, TQueryHelpers = {}>(
    name: string,
    schema?: Schema<T, U, TQueryHelpers>,
    collection?: string,
    skipInit?: boolean
  ): U;
```

I use it by adding <User, Model<User>> to the Schema and to model :

```
const schema = new Schema<User, IUserModel>({
  email: { type: String, required: true },
  password: { type: String, required: true },
})
export const UserModel = model<User, IUserModel>('User', schema, "users")

```

**Comment by traverse1984** [comment]
> A solution while wainting for the PR #10801 to be merged is to use the second declaration of `model` function, declared at line 106 of index.d.ts :

Yeah - comment above the one you upvoted points this out too, it was what let me to line 103.

**Comment by mrfy** [comment]
UP!
It is very frustrating.

**Comment by davecaruso** [comment]
the pr fix doesnt seem to work to help performance. maybe it did, but its still slow. i stumbled upon this once trying the library and finding the types to be slow. the second overload on model() does solves the issue though.

**Comment by traverse1984** [comment]
> the pr fix doesnt seem to work to help performance. maybe it did, but its still slow. i stumbled upon this once trying the library and finding the types to be slow. the second overload on model() does solves the issue though.

Can you confirm with certainty that you’re using the version with that change? For me, the slowdown seemed to relate specifically to Schema<any>. What kind of performance delay are you seeing?

**Comment by yarabramasta** [comment]
This is also happen to me.  Like this validate below.

Or you can check it with other mongoose function like ```schema.pre()```.

```typescript
const schema = new mongoose.Schema({
  email: {
    type: String,
    unique: true,
    required: "Email address is required",
    validate: [validateEmail, "Email address is not valid"],
  },
});
```

I'am using mongoose v6.0.9.

**Comment by haouarihk** [comment]
just don't generate types on the fly .. just make it save them somewhere on the module.. maybe make a third party tool like the prisma one that generate d.ts types for every table once (or whenever the user changes the tables)
most developers don't change their schema that much.. it's just frustrating that types don't update fast enough even if i'm not changing any schema

**Comment by Speckope** [comment]
Hey, I copied the example from https://mongoosejs.com/docs/typescript.html and VSCode started lagging extremely. After commenting out schema definition, lagging stops.
"mongoose": "^6.0.11",
node: v14.18.1 
Linux: Ubuntu 20.04

**Comment by Shumuu** [comment]
> Am I the only one who gets problem this with Javascript too, not just Typescript?

Well, that would be weird as it seems this issue is related to the TypeScript types that take too long to resolve properly. Could you share a barebones repo?

**Comment by andreyvital** [comment]
@lorand-horvath @gregorynicholas AFAIK you can't re-publish a version that's already published, so it's not even doable on NPM side

**Comment by ullaskunder3** [comment]
Update I tried some of the mongoose version installing, checking performance & 🔁 
mongoose -v
- @5.11.10 works with seconds of delay 1 critical severity vulnerability
- @5.12.0 works perfectly but with 1 critical severity vulnerability
- @5.12.10 same issue slow `loading`
- @5.13.0 same issue slow `loading`
- @5.13.5 same issue slow `loading`
- @6.0.0 works fine but with 1 critical vulnerabilities
- @6.0.10 slow down TS/JS performance
- @6.0.5 🥇 work awesome with 0 vulnerabilities  ✔️

**Comment by traverse1984** [comment]
Overall this seems like a real head scratcher as everyone seems to be reporting something a bit different.

**Comment by ShadiestGoat** [comment]
Yeah! I remember from my testing, 5.11 was slow as well? Heh might be wrong tho idk

**Comment by realnatuki** [comment]
> Really? For me it slows all content, not just mongoose related...

Same it slows everything down, been busy for days uninstalling and reinstalling code and extensions.

**Comment by realnatuki** [comment]
> UPDATE Well downgrading works fine... 5.10.19

Yes it's way better fixes slowing the code down but comes with 2 vulnerabilities that require update to version 6

**Comment by cainaf** [comment]
i did test the versions as @ullaskunder3 suggested

@6.0.5 is performant, but with some typing errors
@6.0.10 fixed those errors, but is back to being super slow

**Comment by ShadiestGoat** [comment]
Heya, @vkarpov15 , I have a idea. 
I'm quite busy atm, but if you could take a look at [my module](https://github.com/ShadiestGoat/mongosteel), which shares mongoose's methadology, and check out the types. While not perfect, and also slowing down to some extent, they are a lot faster compared to mongoose's, though it's possible that it's because theres just less of them. You might want to check them out to see if theyd fit in with mongoose's, and I think that at least some portion would (excludiong the similar parts haha)

**Comment by andreialecu** [comment]
https://github.com/microsoft/TypeScript/pull/46429

It appears that a performance issue upstream in TypeScript has recently been released as part of 4.5.

I haven't been able to test it yet but the issues linked to the PR are extremely familiar with the mongoose degradation.

**Comment by ssipos90** [comment]
Latest tsserver uses ts 4.5.2 as per package.json and there's no joy.

**Comment by skhilliard** [comment]
Sorry for taking so long to provide this extra bit of information about the out of memory issue I mention above (https://github.com/Automattic/mongoose/issues/10349#issuecomment-953125529) but I was in the middle of moving...
[SampleCode.zip](https://github.com/Automattic/mongoose/files/7637560/SampleCode.zip)

In the ZIP file are two folders...testbefore uses mongoose v5.12.8. testafter uses mongoose v5.13.13.

You can see the dramatic difference in memory usage by running a local build in VSCode. Navigate to the service folder, run npm ci to install the packages, and then run 'npm run build'. You can look at the memory usage of node.exe and see that the amount of memory used in the after version is dramatically higher.

**Comment by anthonypenna** [comment]
I'm suffering from this as well; Mongoose type definitions make the Typescript Language Server extremely slow.

**Comment by ssipos90** [comment]
@anthonypenna suffering is the correct word. I started working without lsp and watching the terminal for errors.

**Comment by shahidcodes** [comment]
We're all suffering, I am using `"mongoose": "^6.0.13"`, nothing just a single schema it's painfully slow. Maybe I should switch to plain MongoDB and typescript

**Comment by smolinari** [comment]
My project is on the newest version of Mongoose, and there is still a bit of a lag, but it isn't anywhere near as bad as it used to be. That is currently my experience. 

Scott

**Comment by honguyenminh** [comment]
6.0.5 works perfectly fine, while 6.1.2 is lagging really hard. Not sure what changed that made it so slow.

**Comment by barca-reddit** [comment]
I am using mongoose@6.1.3 on Ubuntu 20.04 running VSCode 6.13.2 and I can confirm I am also experiencing a noticeable and unpleasant delay, at least about 2 seconds of "lag", sometimes more.

Using the [example schema](https://mongoosejs.com/docs/typescript.html#creating-your-first-document) provided in Mongoose docs for Typescript makes working with the whole file really slow, I am talking about Intellisense, code-completions, error checking etc. And if I save this file to let's say `User.ts` and import it anywhere else, it would affect working with that other file as well.

**User.ts**
```ts
import { Schema, model } from 'mongoose';

interface User {
    name: string;
    email: string;
    avatar?: string;
}

const schema = new Schema<User>({
    name: { type: String, required: true },
    email: { type: String, required: true },
    avatar: String
});

export const UserModel = model<User>('User', schema);

some invalid code // <= even basic error checking becomes slow
```
![image](https://user-images.githubusercontent.com/35380464/147495463-26195cfb-ace2-45e8-877a-bb07873b7d0e.png)

This can take up to 2-3-4 seconds sometimes.

**SomeOtherFile.ts**
```ts
import { UserModel } from './User';
console.log('hello world');

/* This works all fine and near instantaneous as usual,
    but as soon as we uncomment the code below,
    working with this particular file becomes laggy */

// const u = new UserModel();
```

**Comment by vieuxtemps** [comment]
I can also confirm that 6.0.5 has near instant autocomplete on vscode while not introducing any breaking changes to my project (downgraded from 6.0.15).

**Comment by vkarpov15** [comment]
We did some more digging and found a big improvement today. For the below script:

```ts
import { Schema, Model, model } from "mongoose";
  
interface User {
   name: string;
   email: string;
   avatar?: string;
}

interface UserModel extends Model<User> {
   fetchUser(name: string): Promise<User>;
}

const schema = new Schema<User>({
   name: { type: String, required: true },
   email: { type: String, required: true },
   avatar: String,
});

const User = model<User, UserModel>("User", schema);
```

With Mongoose 6.1.4, here's the extended diagnostics output:

```
Instantiations:             177821
Memory used:               167866K
Check time:                  2.87s
Total time:                  3.96s
```

With simplifying the `Schema` class and avoiding using `ExtractQueryHelpers` and instantiating `HydratedDocument` in `Schema`, we get the below output:

```
Instantiations:              44568
Memory used:               135159K
Check time:                  1.99s
Total time:                  3.11s
```

There's some minor upgrade friction in that you now would need to specify `TQueryHelpers` as a 4th generic to `Schema` for type checking, but I think the perf improvement is worth it.

**Comment by qualphi-master** [comment]
> Same as @ShadiestGoat for me. Even `eslint` gets very slow. If I delete `index.d.ts` from mongoose lib inside node_modules, performance of everything becomes instant again.

Did as he did and All works well now! Wonder what went wrong

**Comment by vkarpov15** [comment]
So right now the most meaningful improvement I've been able to find is removing [`SchemaDefinitionType`](https://github.com/Automattic/mongoose/blob/a416f6a95b449c65d7cc1ba68483a808c12908c3/types/index.d.ts#L2106). That further reduces instantiations from 53k to 49k.

I'm going to close this issue. https://github.com/Automattic/mongoose/commit/1141ec487b12376b2eb7ccf0123a1f6d1245a149 seems to have solved the performance issues that most people on this thread have mentioned. We've kept this open as a way to track our overall efforts to support TypeScript performance, but I think the biggest thing we can do for TypeScript performance now is #11615.

**Comment by skhilliard** [comment]
@vkarpov15 Any chance that these recent commits for performance improvements can be applied to the 5.x branch?

Thanks,
Kelly

**Comment by lorand-horvath** [comment]
@skhilliard @vkarpov15 I've already asked this a while back in https://github.com/Automattic/mongoose/issues/10349#issuecomment-947433549 but I highly doubt it. For all intents and purposes, mongoose v5 seems to be abandoned.

**Comment by Uzlopak** [comment]
Well, I can't speak for vkarpov but I will not invest time for mongoose 5. The question is more if vkarpov is willing to accept PRs for mongoose 5 and if you want to create PRs for v5. If vkarpov accepts PRs for v5 you could provide PRs with your desired performance changes.

**Comment by vkarpov15** [comment]
We'll see what we can do for 5.x.

**Comment by pouriamoosavi** [comment]
I have `mongoose` on both typescript and javascript projects. In both it is slow after 5.11. But in typescript, VSCode also  shows me an error:
```
Type instantiation is excessively deep and possibly infinite
```
I thought maybe this would help to find the problem.

**Comment by Uzlopak** [comment]
Use latest version 6 and latest version of vscode or force vscode to use latest typescript if you want to use older vscode. Latest vscode versions have better typings performance.

**Comment by lorand-horvath** [comment]
@Uzlopak lol - latest VSCode actually have terrible "performance" https://github.com/microsoft/vscode/issues/146737

**Comment by Uzlopak** [comment]
Or forcde vscode to use latest typescript.

**Comment by cgarrovillo** [comment]
I have the opposite experience.  VSCode February 2022 (TS >4.6) has terrible performance, where as VSCode January 2022 (TS <4.6) shows no regression.

For now, my VSCode is March 2022, while my Intellisense server for typescript is pointed to a global install of TypeScript 4.5.5

**Comment by Zealotmidnight** [comment]
> I have the opposite experience. VSCode February 2022 (TS >4.6) has terrible performance, where as VSCode January 2022 (TS <4.6) shows no regression.
> 
> For now, my VSCode is March 2022, while my Intellisense server for typescript is pointed to an install of `TypeScript 4.5.5`

have the same problem can't do anything to fix it, so I move on to mongodb native. hope everything get better soon

ps. sorry for my English

**Comment by andrewbranch** [comment]
Can anyone share an open source repo that has a lot of mongoose usage and seems affected by this issue? I’m hoping to try some experiments and see how they impact a codebase larger than the tiny example repro shared at the top of this thread.

**Comment by SethWen** [comment]
Same behavior for me.

### 1. my environment
```
nodejs: 12.20.2  
tsc: 4.2.4  
mongoose: 5.11.15
```

### 2. reproduce
open the code in vscode
```
import { Schema, Model, Document, FilterQuery, UpdateQuery } from 'mongoose';
import * as assert from 'assert';

const mySchema = new Schema({
    rid: { type: String },
    tag: { type: String },
    type: { type: String }, 
    size: { type: String },
    colorMode: { type: String },
    dimension: { type: String },
    uploadTime: { type: Date },
    name: { type: String },
    ctime: { type: Date },
    utime: { type: Date },
});
```

![](https://modorone-public.oss-cn-beijing.aliyuncs.com/pictures/bug-report.jpg)

Is there any solution now?

**Comment by andrewbranch** [comment]
@SethWen your version of TypeScript is over a year old.

**Comment by dtboy1995** [comment]
Any updates？

**Comment by vkarpov15** [comment]
@dtboy1995 this issue is fixed as far as we know in 6.x. Please open a new issue and follow the issue template.

**Comment by andrewbranch** [comment]
@vkarpov15 would you be able to point me to any PRs related to that perf improvement, or notes from the investigation/fix? I’ve been looking at the complaints we’ve seen over at TypeScript / VS Code and it does look like all of those folks were using mongoose 5.x. Trying to determine if there’s anything actionable for us, or if the v5 types were just inherently inefficient.

**Comment by andreialecu** [comment]
@andrewbranch, take a look at these commits: https://github.com/Automattic/mongoose/search?q=10349&type=commits

I think most of the perf changes made the types worse, removing features to improve performance.

**Comment by andrewbranch** [comment]
Using the repro at https://github.com/Automattic/mongoose/issues/10349#issuecomment-860981794

mongoose@5.13.14:

```
$ npx tsc --extendedDiagnostics

Files:                         113
Lines of Library:            28686
Lines of Definitions:        45982
Lines of TypeScript:            18
Lines of JavaScript:             0
Lines of JSON:                   0
Lines of Other:                  0
Nodes of Library:           122244
Nodes of Definitions:       121403
Nodes of TypeScript:            66
Nodes of JavaScript:             0
Nodes of JSON:                   0
Nodes of Other:                  0
Identifiers:                 86420
Symbols:                    133695
Types:                       58605
Instantiations:             148743
Memory used:               171707K
Assignability cache size:    20368
Identity cache size:           133
Subtype cache size:              0
Strict subtype cache size:       0
I/O Read time:               0.19s
Parse time:                  0.68s
ResolveModule time:          0.06s
ResolveTypeReference time:   0.00s
Program time:                0.99s
Bind time:                   0.29s
Check time:                  2.56s
printTime time:              0.00s
Emit time:                   0.00s
Total time:                  3.84s
```

mongoose@6.4.0:

```
Files:                         140
Lines of Library:            28686
Lines of Definitions:        52590
Lines of TypeScript:            18
Lines of JavaScript:             0
Lines of JSON:                   0
Lines of Other:                  0
Nodes of Library:           122244
Nodes of Definitions:       141748
Nodes of TypeScript:            66
Nodes of JavaScript:             0
Nodes of JSON:                   0
Nodes of Other:                  0
Identifiers:                 93349
Symbols:                    227668
Types:                      106226
Instantiations:             532762
Memory used:               250874K
Assignability cache size:    34548
Identity cache size:          9758
Subtype cache size:              0
Strict subtype cache size:       0
I/O Read time:               0.22s
Parse time:                  0.74s
ResolveModule time:          0.05s
ResolveTypeReference time:   0.02s
Program time:                1.09s
Bind time:                   0.32s
Check time:                  4.18s
printTime time:              0.00s
Emit time:                   0.00s
Total time:                  5.59s
```

v6 seems to be significantly slower, at least on this base case 🤔

**Comment by andrewbranch** [comment]
No, that’s just the difference in lib files in node_modules caused by upgrading mongoose.

**Comment by andreialecu** [comment]
@andrewbranch, to completely understand this issue, you could also try benchmarking `mongoose@~5.10` with the types from DefinitelyTyped (`@types/mongoose`). 

Mongoose added official/embedded types in `5.11`, but it was a completely different implementation than DT ones. 

The types from DT did not suffer from any performance issues, from what I remember.

**Comment by andrewbranch** [comment]
I was able to knock 1s off the v6 total time via the following process:

1. `tsc --generateTrace .`
2. `npx analyze-trace .`
3. Find the highest line that says `Determine variance of type ...`
4. Look up what type the id from (3) was referring to in types.json
5. Add [variance annotations](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#optional-variance-annotations-for-type-parameters) on that type (take a guess at the correct annotation and TS will tell you if you’re wrong—everything was either `out` [covariant] or `in out` [invariant])
6. Repeat

**Comment by vkarpov15** [comment]
@andrewbranch can you please paste the exact Mongoose script you're using? Our TypeScript benchmarks show significantly lower instantiations, memory usage, and total time on 6.4. For example: https://github.com/Automattic/mongoose/runs/7050913582?check_suite_focus=true

Here's the script our TS benchmark uses: https://github.com/Automattic/mongoose/blob/3a4ba27a75622a9fc370683744b8822f3f9dbfdf/benchmarks/typescript/simple/index.ts

**Comment by andrewbranch** [comment]
@vkarpov15 

tsc: 4.7.4

tsconfig.json:

```json
{
  "compilerOptions": {
    "target": "esnext",
    "module": "commonjs",
    "strict": true,
    "noEmit": true
  }
}
```

index.ts:

```ts
import { model, Model, Schema } from "mongoose";

export interface UserSH {
    username: string,
    password: string,
    maxMb: number,
    submitted: string[],
    id: string
}

export const UserSchema = new Schema<UserSH, Model<UserSH>, UserSH>({
    id: "string",
    maxMb: "number",
    username: "string",
    submitted: ["string"],
    password: "string"
})
```

**Comment by JuanValdivieso** [comment]
I can confirm the issue is still present on Mongoose v6.4.1 & VS Code 1.68.1. I'm currently following a very simple tutorial and requiring Mongoose slows down IntelliSense tremendously. Commenting out all imports restores performance to its normal level.

**Comment by Chudroy** [comment]
I also have this problem, the last version of mongoose that is fast for me is 5.10.4

**Comment by sandraroz** [comment]
Happy to find this thread. I thought my PC was starting to die.

Same slow behaviour found on MAC and WINDOWS machines. I noticed the problem while doing tutorials to learn mongoose and somehow ended up on this thread.

Thank you @Chudroy for finding a mongoose version that is still fast. 5.10.4 is indeed not slowing down intellisense with the latest version of vscode (1.68.1).

**Comment by sandessat** [comment]
This issue needs to be addressed urgently, it's almost impossible to program with a delayed appearing completion menu. On the other hand, it is not my preference to have to downgrade to solve this problem, as I want to keep everything up to date to avoid problems, among others.

**Comment by ssipos90** [comment]
@sandraroz it doesn't matter, your IDE still uses libs with d.ts for completion.

**Comment by geauser** [comment]
I tried to use @Thisura98 fix and it stopped lagging, but now I lost type checking and autocomplete on my schemas... Using tsc v4.7.4 and VScode 1.69.0

**Comment by Near99** [comment]
> I tried to use @Thisura98 fix and it stopped lagging, but now I lost type checking and autocomplete on my schemas... Using tsc v4.7.4 and VScode 1.69.0

Same here. The issue really needs to be addressed urgently.

**Comment by ismm0101** [comment]
Any solutions yet?

**Comment by LeBogoo** [comment]
> I am experiencing the problem since Mongoose has adopted it's own official type definition. I have a workaround that is to delete `index.d.ts` of mongoose in node_module and then `npm i -D -E @types/mongoose@5.10.5` to temporarily use the DefinitelyTyped type definition.

This works for me! Unfortunately it's only a temporary workaround because it's not stored in package.json :/

**Comment by Alexandre-Fernandez** [comment]
Same problem on linux

**Comment by GalaxyVinci05** [comment]
I'm also having the same issue since I started rewriting my project in TypeScript, and @Chudroy 's solution seems to be the best one
`npm i mongoose@5.10.4`
`npm i @types/mongoose@5.10.4`
seems to be working fine for me

**Comment by ismailelbadawy** [comment]
I'm also experiencing the same problem

**Comment by baa76** [comment]
> Same as @ShadiestGoat for me. Even `eslint` gets very slow. If I delete `index.d.ts` from mongoose lib inside node_modules, performance of everything becomes instant again.

Same issues - the quoted fix returned performance to normal.

**Comment by AdsonCicilioti** [comment]
Same problem here. VS Code Very Slow with typescript project.

**Comment by Alexandre-Fernandez** [comment]
@Chudroy's also worked for me, you just need to add `extends Document` to your TS schema interfaces.

**Comment by grantjayy** [comment]
As of now is the only solution to downgrade to 5.10.19 and install DefinetlyTyped types ?

This is a big issue. It's completely unbearable to code with the lag.

Are there any updates to this being fixed ?

**Comment by vkarpov15** [comment]
@grantjayy it is fixed in Mongoose 6. Is there anything blocking you from upgrading?

**Comment by andrewbranch** [comment]
@vkarpov15 in my experiments, Mongoose 6 was slower. I offered detailed data starting at https://github.com/Automattic/mongoose/issues/10349#issuecomment-1164648764. Many others since have anecdotally confirmed my results. As I said, I tested the same source as your benchmark and observed an order of magnitude higher instantiations and type count. I think something must be misconfigured in your benchmark—I couldn’t find where it’s instrumented.

**Comment by eroglu94** [comment]
@vkarpov15 currenly using `"mongoose": "^6.4.6",` and it is slowing down the suggestions a lot. When i comment out schema definition, intellisense is become faster again.

**Comment by sandessat** [comment]
I'm curious if this lag issue happens too with Flow... If someone have tried using Mongoose with Flow, I would really like to know how was your code editor experience, and if it lagged, or not.

**Comment by GalaxyVinci05** [comment]
> I'm also having the same issue since I started rewriting my project in TypeScript, and @Chudroy 's solution seems to be the best one `npm i mongoose@5.10.4` `npm i @types/mongoose@5.10.4` seems to be working fine for me

Just realized that when compiling TypeScript, it tells you to install types for the mongodb package, but when you do, a lot of errors show up at the next compilation.
I installed `@types/mongodb@4.0.7` (latest version), does someone know a version that works?

**Comment by Uzlopak** [comment]
What happens if you patch your package.json like this and do a npm install?
```
"@types/mongodb": "npm:mongodb",
```

**Comment by GalaxyVinci05** [comment]
@Uzlopak this also gives a bunch of errors, it looks like that version isn't compatible with `mongoose@5.10.4` (?)

**Comment by Uzlopak** [comment]
Well you are basically trying to use old typings. Maybe it is better to not use typings at all.

**Comment by GalaxyVinci05** [comment]
That's what I wanted to do, but when I compiled it said that type declaration was missing

**Comment by vkarpov15** [comment]
Confirmed, it looks like some recent changes, including #11650, caused a slowdown that our benchmarks are not picking up on surprisingly. Running `tsc --extendedDiagnostics` from the Mongoose repo's `benchmarks/typescript/simple` repo is still fast, but once you `npm install mongoose` in a separate repo then `tsc --extendedDiagnostics` becomes very slow.

We undid some of the changes from #11650, and we'll figure out what other issues caused perf degradations.

**Comment by nidhalmessaoudi** [comment]
Still facing the same problem here even after upgrading mongoose to the latest version (v6.5.0) from (v6.4.6). Very slow intellisense and auto-completion in VS Code only in a project when mongoose is being used!

**Comment by Uzlopak** [comment]
@mohammad0-0ahmad 

Maybe this is worth the read 
https://effectivetypescript.com/2022/07/30/treemap-for-source-files/

**Comment by vkarpov15** [comment]
@Uzlopak not really. That article's only about detecting unexpected large TS files, which is nice but typically not the cause of TypeScript performance issues.

**Comment by vkarpov15** [comment]
I've confirmed that the cause of this slowdown is #11563. It looks like the cause is that #11563 makes heavy use of `infer`, which can be extremely slow. Here's `tsc --extendedDiagnostics` with 77d2c70:

```
Instantiations:             392799
Memory used:               213855K
Assignability cache size:    29100
Check time:                  2.47s
Total time:                  3.25s
```

Now if I replace `ObtainSchemaGeneric` with the below:

```
type ObtainSchemaGeneric<TSchema, alias extends 'EnforcedDocType' | 'M' | 'TInstanceMethods' | 'TQueryHelpers' | 'TVirtuals' | 'TStaticMethods' | 'TPathTypeKey' | 'DocType'> = any;
```

I get:

```
Instantiations:              60772
Memory used:               163552K
Assignability cache size:    16624
Check time:                  1.67s
Total time:                  2.44s
```

We'll do some more digging to see what we can do.

@mohammad0-0ahmad do you have any ideas how we might make your auto typed schemas code work without relying on `infer`?

**Comment by mohammad0-0ahmad** [comment]
@vkarpov15  @Uzlopak, when I created that PR, I've to test multiple ways to get the best performance without have a lot of changes on the types and prevent as much as possible of regressions. I thought by obtaining the correct doc type  from the beginning "when declaring the schema" might allow us to refactor the entire ts code to avoid unnecessary operations that already exist to get the correct doc type, like lean required_id and so on. 
I am not really sure if that will make things better, I wonder if we should to rely on a specific field or benchmark to measure the changing of the performance like total time, check time or maybe someone else, I couldn't find full documentation on all extendedDiagnostics options fields.

**Comment by mohammad0-0ahmad** [comment]
With other words, I wonder if there is a field that represent the actual waiting time to show intellisense message "or suggestions fields" that make user feel that it is slow. Even if that field affects by end user pc specs, we can maybe specify a custom memory, CPU or cache configuration in a typescript benchmark workflow action file "if that possible".

**Comment by mohammad0-0ahmad** [comment]
I've noticed as well that the current test file that we have rely on manual typed schema, not the auto one which give even higher Instantiations result. So I think we might need as well to create a separate one for the auto typed one to be able to target the different approaches while make any change on the types.

**Comment by mohammad0-0ahmad** [comment]
Something like:
```diff
- const schema = new Schema<User>({
+ const schema = new Schema({
  name: { type: String, required: true },
  email: { type: String, required: true },
  avatar: String
});

- const UserModel = model<User, UserModelInterface>('User', schema);
+ const UserModel = model('User', schema);
```

**Comment by Uzlopak** [comment]
Sry, misclicked with my mobile.

Well tbh I am still kind of bummed since the heavy regressions i caused with the aggregation pipeline typings. But I actually was quite happy about the changes we made regarding the typings. We separated the typings in multiple logical units and tbh. some of the typings are kind of simple. So we can actually kind of say that some type files should be not critical regarding performance (simple interfaces without any special typescript logic). Maybe we should focus on further separating remaining index.d.ts till we just import typings from other files. Then we optimize one schema after the other.

**Comment by mohammad0-0ahmad** [comment]
> I've confirmed that the cause of this slowdown is #11563. It looks like the cause is that #11563 makes heavy use of `infer`, which can be extremely slow. Here's `tsc --extendedDiagnostics` with [77d2c70](https://github.com/Automattic/mongoose/commit/77d2c70c06a94b9084f040ade29e99f1269756c5):
> 
> ```
> Instantiations:             392799
> Memory used:               213855K
> Assignability cache size:    29100
> Check time:                  2.47s
> Total time:                  3.25s
> ```
> 
> Now if I replace `ObtainSchemaGeneric` with the below:
> 
> ```
> type ObtainSchemaGeneric<TSchema, alias extends 'EnforcedDocType' | 'M' | 'TInstanceMethods' | 'TQueryHelpers' | 'TVirtuals' | 'TStaticMethods' | 'TPathTypeKey' | 'DocType'> = any;
> ```
> 
> I get:
> 
> ```
> Instantiations:              60772
> Memory used:               163552K
> Assignability cache size:    16624
> Check time:                  1.67s
> Total time:                  2.44s
> ```
> 
> We'll do some more digging to see what we can do.
> 
> @mohammad0-0ahmad do you have any ideas how we might make your auto typed schemas code work without relying on `infer`?

@vkarpov15  I don't think there is a way to do that without using ```infer```  It might be helpful if there is a way to store the inference result of the scheme to avoid re infer generics multiple time.
It might be helpful to ask someone of TS team like @sandersn or maybe of VSC team @mjbvz.
I hope I didn't bother you by mentioning here.

**Comment by sandersn** [comment]
@andrewbranch has been the TS team member investigating the mongoose types.

**Comment by vkarpov15** [comment]
I did some more digging, haven't made much progress. Managed to make a token improvement by removing one of the `infer` paths in `ObtainSchemaGeneric`, but that doesn't make much of a dent. It does seem that `infer` is necessary here, as @mohammad0-0ahmad indicated, so we'll have to figure out an alternative approach to make tsc not blow up.

**Comment by mrfy** [comment]
@vkarpov15 
In our project we also struggled with the slow performance of TS-server, today we did some tests and disabled all vscode extensions, **performance improved drastically**! In our case, it turned out that one particular [extension ](https://marketplace.visualstudio.com/items?itemName=Vue.vscode-typescript-vue-plugin) affects the TS-server performance, turning it off makes everything run smoothly. Please check on your enviroment

**Comment by vkarpov15** [comment]
I'm gonna close this because of #12284 . We'll do some separate work to improve our TS benchmarks with #12320.

If you're running into a similar issue, please create a standalone repro script that demonstrates slow performance using `tsc --extendedDiagnostics`. It is practically impossible for us to debug "my VSCode is slow", we have no experience with VSCode internals. So unless you can demonstrate slow TypeScript performance, there isn't much we can do unfortunately.
